packages/stryker/src/ScoreResultCalculator.ts

```diff
@@ -0,0 +1,125 @@
+import { MutantResult, MutantStatus, ScoreResult } from 'stryker-api/report';
+import { freezeRecursively } from './utils/objectUtils';
+import * as _ from 'lodash';
+import * as path from 'path';
+
+export default class ScoreResultCalculator {
+
+  static calculate(results: MutantResult[], basePath: string = ''): ScoreResult {
+    const name = this.determineCommonBasePath(results, basePath);
+    const childResults = this.calculateChildScores(results, name, basePath);
+    const numbers = this.calculateNumbers(results, childResults);
+    return freezeRecursively(_.assign(numbers, { name, childResults }));
+  }
+
+  private static compareScoreResults(a: ScoreResult, b: ScoreResult) {
+    const sortValue = (scoreResult: ScoreResult) => {
+      // Directories first
+      if (scoreResult.childResults.length) {
+        return `0${scoreResult.name}`;
+      } else {
+        return `1${scoreResult.name}`;
+      }
+    };
+    return sortValue(a).localeCompare(sortValue(b));
+  }
+
+  private static calculateChildScores(results: MutantResult[], name: string, basePath: string) {
+    const childrenBasePath = path.join(basePath, name) + path.sep;
+    const resultsByFiles = _.groupBy(results, result => result.sourceFilePath.substr(childrenBasePath.length));
+    const uniqueFiles = Object.keys(resultsByFiles);
+    if (uniqueFiles.length > 1) {
+      const filesGroupedByDirectory = _.groupBy(uniqueFiles, file => file.split(path.sep)[0]);
+      return Object.keys(filesGroupedByDirectory)
+        .map(directory => this.calculate(_.flatMap(filesGroupedByDirectory[directory], file => resultsByFiles[file]), childrenBasePath))
+        .sort(this.compareScoreResults);
+    } else {
+      return [];
+    }
+  }
+
+  private static determineCommonBasePath(results: MutantResult[], basePath: string) {
+    const uniqueFiles = _.uniq(results.map(result => result.sourceFilePath));
+    const uniqueFileDirectories = uniqueFiles.map(file => file.substr(basePath.length).split(path.sep));
+    if (uniqueFileDirectories.length) {
+      return uniqueFileDirectories
+        .reduce((previousDirectories, currentDirectories) => previousDirectories.filter((token, index) => currentDirectories[index] === token))
+        .join(path.sep);
+    } else {
+      return '/';
+    }
+  }
+
+  private static calculateNumbers(mutantResults: MutantResult[], childScores: ScoreResult[]) {
+    if (childScores.length) {
```

Comment:
> If we would pass in a MutationScoreResult interface for the MutationScoreResult that we're working on instead of childScores, we could just check if the current MutationScoreResult is a directory.

No we cannot. The `ScoreResult` interface is entirely `readonly` (something we should also add to other reporter interfaces). So after instantiating, we cannot alter the properties anymore. We could of course copy the values to a new instance, but i actually like it this way.

> I would also suggest moving the check to the calculate method.

I wouldn't do that. I think the `if` statement represents a new layer of abstraction, which i like to add to it's own function.

Note. We could remove the `calculateNumbersBasedOnChildScores` method all together. It's a performance improvement.